printers/vcd_witness_printer.cpp

```diff
@@ -0,0 +1,571 @@
+  } while (pos < str.length() && prev < str.length());
+  return tokens;
+}
+
+
+/// convert a widith to a verilog string
+static std::string width2range(uint64_t w) {
+  if (w > 1)
+    return std::string("[") + std::to_string(w - 1) + ":0]";
+  return "";
+}
+
+/// copied from btor2_witness_printer,
+/// so that we don't need to include
+/// because its header contains also 
+/// implementation, this creates troubles
+static std::string as_bits(std::string val)
+{
+  // TODO: this makes assumptions on format of value from boolector
+  //       to support other solvers, we need to be more general
+  std::string res = val;
+
+  if (val.length() < 2) {
+    throw CosaException("Don't know how to interpret value: " + val);
+  }
+
+  if (res.substr(0, 2) == "#b") {
+    // #b prefix -> b
+    res = res.substr(1, val.length() - 1);
+  } else if (res.substr(0, 2) == "#x") {
+    throw CosaException("Not supporting hexadecimal format yet.");
+  } else {
+    res = res.substr(5, res.length() - 5);
+    std::istringstream iss(res);
+    std::vector<std::string> tokens(std::istream_iterator<std::string>{ iss },
+                                    std::istream_iterator<std::string>());
+
+    if (tokens.size() != 2) {
+      throw CosaException("Failed to interpret " + val);
+    }
+
+    res = tokens[0];
+    // get rid of ")"
+    std::string width_str = tokens[1].substr(0, tokens[1].length() - 1);
+    size_t width = std::stoull(width_str);
+    mpz_class cval(res);
+    res = cval.get_str(2);
+    size_t strlen = res.length();
+
+    if (strlen < width) {
+      // pad with zeros
+      res = std::string(width - strlen, '0') + res;
+    } else if (strlen > width) {
+      // remove prepended zeros
+      res = res.erase(0, strlen - width);
+    }
+    return res;
+  }
+  return res;
+}
+
+
+static std::string as_decimal(std::string val)
+{
+  // TODO: this makes assumptions on format of value from boolector
+  //       to support other solvers, we need to be more general
+  std::string res = val;
+
+  if (val.length() < 2) {
+    throw CosaException("Don't know how to interpret value: " + val);
+  }
+
+  if (res.substr(0, 2) == "#b") {
+    // #b prefix -> b
+    res = res.substr(2, val.length() - 1);
+    mpz_class cval(res, 2);
+    res = cval.get_str(10);
+  } else if (res.substr(0, 2) == "#x") {
+    throw CosaException("Not supporting hexadecimal format yet.");
+  } else {
+    res = res.substr(5, res.length() - 5);
+    std::istringstream iss(res);
+    std::vector<std::string> tokens(std::istream_iterator<std::string>{ iss },
+                                    std::istream_iterator<std::string>());
+
+    if (tokens.size() != 2) {
+      throw CosaException("Failed to interpret " + val);
+    }
+
+    res = tokens[0];
+    // get rid of ")"
+    std::string width_str = tokens[1].substr(0, tokens[1].length() - 1);
+    size_t width = std::stoull(width_str);
+    mpz_class cval(res);
+    res = cval.get_str(10);
+    // no need for padding
```

Comment:
You can just delete this commented out code if it's not needed.

Reply:
For point 2, it was included in 06d7b6b174c1f2c4097102d4f8841cb575afbb50. However, that commit also includes "naming the terms if they have names" which you already addressed by 823d9582bbf951ff0144203153558134ab1cfc79

Let me create a commit based on the master branch as the reference, so it will be clear what I changed. Basically, my solution to point 2 was to create a preprocessing step, because the name of the state (that goes with the output) comes later than the state declaration, and the preprocessing step was just to collect the names of the outputs.

Will post it here once I finish. Thanks!

Reply:
Here is the change: 6ae74c9eb1cd6ae771669e4f95453e80376bca9f
If you think this is okay, I will merge it to this pulling request.

Reply:
Looks good to me, thanks! Just one question, what is “sth”? https://github.com/upscale-project/cosa2/blob/6ae74c9eb1cd6ae771669e4f95453e80376bca9f/frontends/btor2_encoder.cpp#L183

Reply:
sth. = something  :-)